### ISSUE TITLE
Avoiding nesting like pre > p or p > h2 that works but aren't valid in html

### ISSUE THREAD

**Issue by caub** [w3c/editing#115]
With Chrome you can [end up](https://bugs.chromium.org/p/chromium/issues/detail?id=582020) with such nesting, not only they are not valid in HTML, but you can't un-nest the blocks.

Firefox maintains a low depth of blocks, I believe it's a good approach

There are no specs, but are what are the thoughts on this?

**Comment by johanneswilm** [comment]
What do you mean by "You can't un-nest the blocks", @cauburtin ? Do you mean the editor of a document using a raw contenteditable form? Or do you mean it's impossible to do this in JavaScript?

**Comment by caub** [comment]
It's impossible to unnest this invalid html tree with all the possible commands (trying formatBlock 'p', or 'div', etc..)

you can also see some p > p on Chrome

**Comment by johanneswilm** [comment]
@cauburtin: Ok, but this is only if you limit yourself to only use execCommand, right? There shouldn't be any problem if you use JavaScript commands related to creating, deleting and moving dom nodes around, right?

**Comment by caub** [comment]
yes, but it involves a lot of parsing rules, and a separate undo manager

**Comment by johanneswilm** [comment]
@cauburtin Good point about undo manager. Feel free to propose concrete changes to the execCommand and ContentEditableTrue specs. As long as you only propose default behavior, I would support that. But be aware that browsers don't expect to follow the relevant specs any time in the next few years if ever. For now and the next few years the only option seems to be to maintain a separate undo manager.

**Comment by caub** [comment]
@johanneswilm thanks I'm quite happy with execCommand

one thing that I'd like being changed or spec'ed is: 
inserting an image in a link, remove the link, but it shouldn't a link can perfectly contain an image, it's useful for 'lightbox' effects: issue: http://jsfiddle.net/crl/c5m94qcd/11/ https://bugs.chromium.org/p/chromium/issues/detail?id=590595 (note in works on Firefox)

**Comment by johanneswilm** [comment]
Could you write a PR for your proposal? As we have limited resources, I'm trying not to spend it on something that isn't prioritized by browsers currently.

**Comment by caub** [comment]
ok no problem, oh I realize insertImage is considered as a 'block' command (but `<img>` are inline-block) that's maybe why the old spec didn't allow `<img>` inside `<a>`

hmm I don't know where to put "preserve a parent `<a>` when inserting an image inside it" in  https://w3c.github.io/editing/execCommand.html#the-insertimage-command or insertHTML just above

**Comment by johanneswilm** [comment]
I don't think formalities matter much for now, as the spec draft isn't compliant anyway. Just make sure it's very clear what you are proposing. if your proposal is about adopting the existing behavior of one browser (Firefox or otherwise), I would think it will be more likely there won't be opposition to it.
